### ISSUE TITLE
CMake subproject does not add -pthread to the linker command line

### ISSUE THREAD

**Issue by DaanDeMeyer** [mesonbuild/meson#6386]
**Describe the bug**

I've been trying to make [reproc](https://github.com/DaanDeMeyer/reproc) useable as a meson subproject. The current issue is that demo executables fail to link due to `-pthread` not being added to the linker command line by meson.

**To Reproduce**

meson.build:

```
project('tutorial', 'c')

cmake = import('cmake')

reproc = cmake.subproject('reproc')
libreproc = reproc.dependency('reproc')

executable('demo', ['main.c'], dependencies: [libreproc])
```

main.c (one of reproc's examples):

```c
#include <reproc/reproc.h>

#include <stdio.h>
#include <stdlib.h>

// Forwards the provided arguments to `reproc_start` which starts a child
// process that inherits the standard streams of the parent process. It will
// read its input from the stdin of this process and write its output to the
// stdout/stderr of this process.
int main(int argc, const char *argv[])
{
  if (argc <= 1) {
    fprintf(stderr,
            "No arguments provided. Example usage: ./inherit cmake --help");
    return 1;
  }

  reproc_t *process = reproc_new();

  reproc_options options = { .redirect = { .in = REPROC_REDIRECT_INHERIT,
                                           .out = REPROC_REDIRECT_INHERIT,
                                           .err = REPROC_REDIRECT_INHERIT } };
  int r = -1;

  r = reproc_start(process, argv + 1, options);
  if (r < 0) {
    goto cleanup;
  }

  r = reproc_wait(process, REPROC_INFINITE);
  if (r < 0) {
    goto cleanup;
  }

  r = reproc_exit_status(process);

cleanup:
  reproc_destroy(process);

  if (r < 0) {
    fprintf(stderr, "%s\n", reproc_error_string(r));
  }

  return abs(r);
}
```

```sh
git clone https://github.com/DaanDeMeyer/reproc.git subprojects/reproc
meson build
ninja -C build
```

**Expected behavior**

The build succeeds and I can run `build/demo cmake --help`. Instead, linking fails because `pthread_sigmask` cannot be found.

reproc enables `THREADS_PREFER_PTHREAD_FLAG` before calling `find_package(Threads)`. The pthreads dependency is added to reproc with `target_link_libraries(reproc PRIVATE Threads::Threads)`. Changing `PRIVATE` to `PUBLIC` or disabling `THREADS_PREFER_PTHREAD_FLAG` does not fix the issue.

**system parameters**
* Platform: Native
* OS: Arch Linux
* Python 3.8.1
* meson 0.52.1
* ninja 1.9.0

**Comment by mensinda** [comment]
I was able to build your test project with the following `meson.build`:

```meson
project('tutorial', 'c')

cmake = import('cmake')

reproc = cmake.subproject('reproc')
libreproc = reproc.dependency('reproc')
threads = dependency('threads')

executable('demo', ['main.c'], dependencies: [libreproc, threads])
```

Unfortunately, you have to manually add the `threads` dependency, since meson (currently) does not know which flags from `reproc` are required for linking `demo`. Sadly, this is an internal limitation from meson and it is hard to work around this.

**Comment by DaanDeMeyer** [comment]
Alright, should anyone ever ask about using reproc with meson I can point them to this solution.

I assume the issue can be left open since there's still room for improvement?

**Comment by mensinda** [comment]
Yes. I still have to figure out if this is a general problem or if it is pthread specific.

**Comment by mensinda** [comment]
This is currently a general problem with static libraries. Unfortunately, CMake does not specify the required link args for *consuming* static libraries.

**Comment by DaanDeMeyer** [comment]
Yes you're right, passing `-DBUILD_SHARED_LIBS=ON` fixes the issue. I can report this in the CMake repository. Is it specifically the CMake File API that does not report the required link args?

**Comment by mensinda** [comment]
Yes, this is the issue. Now that I think about it, it does make sense in some way that the link flags are not specified since they are not required for *building* the library. CMake does add the flags to an executable when linking with the static library. So, I am not sure if this is accidental or the intended behavior.

I think I can work around this by looking at the trace output and parsing `target_link_libraries`. We already use the CMake trace for `add_custom_command` anyway.

As for reporting this to CMake, I am not sure if they are willing to support this. See the discussion [here](https://gitlab.kitware.com/cmake/cmake/issues/19348). That being said, my [JSON trace MR](https://gitlab.kitware.com/cmake/cmake/merge_requests/4102) was merged recently.

**Comment by mensinda** [comment]
Should be fixed with #6388.
